File: src/utils/session-store.js
"use strict";

const fs = require("fs");
const path = require("path");

// 会话文件放在项目目录下，便于和项目绑定，不污染全局环境。
const SESSION_FILE = path.join(process.cwd(), ".sessions", "session-store.json");

// 读取会话缓存文件。
// 文件不存在或内容异常时，返回空对象，保证主流程不崩溃。
function loadSessionStore() {
  try {
    const raw = fs.readFileSync(SESSION_FILE, "utf8");
    const parsed = JSON.parse(raw);
    return parsed && typeof parsed === "object" ? parsed : {};
  } catch {
    return {};
  }
}

// 把会话缓存写回磁盘；目录不存在时自动创建。
function saveSessionStore(store) {
  const dir = path.dirname(SESSION_FILE);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(SESSION_FILE, JSON.stringify(store, null, 2), "utf8");
}

// 读取某个 provider 的 session id（例如 codex）。
function getSessionId(provider) {
  const store = loadSessionStore();
  return store[provider]?.sessionId || null;
}

// 更新某个 provider 的 session id，并记录更新时间。
function setSessionId(provider, sessionId) {
  const store = loadSessionStore();
  store[provider] = {
    sessionId,
    updatedAt: new Date().toISOString(),
  };
  saveSessionStore(store);
}

module.exports = {
  SESSION_FILE,
  getSessionId,
  setSessionId,
  loadSessionStore,
};

File: src/utils/session_store.py
"""Persistent provider session store utilities."""

from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, Optional

SESSION_FILE = Path.cwd() / ".sessions" / "session-store.json"
DEBUG_ENV = "FRIENDS_BAR_DEBUG"


def _debug_log(message: str) -> None:
    """Print optional debug logs when FRIENDS_BAR_DEBUG is enabled."""
    if os.environ.get(DEBUG_ENV):
        print(f"[session_store] {message}", file=sys.stderr)


def load_session_store() -> Dict[str, Any]:
    """Load full session store from disk."""
    if not SESSION_FILE.exists():
        return {}

    try:
        raw = SESSION_FILE.read_text(encoding="utf-8")
        parsed = json.loads(raw)
    except (OSError, UnicodeDecodeError, json.JSONDecodeError) as exc:
        _debug_log(f"load failed, fallback to empty store: {exc}")
        return {}

    if isinstance(parsed, dict):
        return parsed

    _debug_log("invalid store type, fallback to empty store")
    return {}


def save_session_store(store: Dict[str, Any]) -> None:
    """Atomically save full session store to disk."""
    SESSION_FILE.parent.mkdir(parents=True, exist_ok=True)
    payload = json.dumps(store, indent=2, ensure_ascii=False)
    tmp_file = SESSION_FILE.with_name(
        f".{SESSION_FILE.name}.tmp-{os.getpid()}-{int(datetime.now(timezone.utc).timestamp() * 1_000_000)}"
    )

    try:
        tmp_file.write_text(payload, encoding="utf-8")
        os.replace(tmp_file, SESSION_FILE)
    finally:
        if tmp_file.exists():
            try:
                tmp_file.unlink()
            except OSError as exc:
                _debug_log(f"cleanup temp file failed: {exc}")


def get_session_id(provider: str) -> Optional[str]:
    """Read one provider session id."""
    store = load_session_store()
    value = store.get(provider, {})
    if isinstance(value, dict):
        session_id = value.get("sessionId")
        if isinstance(session_id, str) and session_id.strip():
            return session_id
    return None


def set_session_id(provider: str, session_id: str) -> None:
    """Update one provider session id and save."""
    store = load_session_store()
    store[provider] = {
        "sessionId": session_id,
        "updatedAt": datetime.now(timezone.utc).isoformat(),
    }
    save_session_store(store)


def clear_session_id(provider: str) -> None:
    """Delete one provider session mapping."""
    store = load_session_store()
    if provider in store:
        del store[provider]
        save_session_store(store)
